File: src/main/java/ui/game/components/SquarePane.java
package ui.game.components;

import javafx.beans.value.ObservableNumberValue;
import javafx.scene.paint.Color;
import javafx.scene.paint.Paint;
import javafx.scene.shape.Rectangle;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import ui.game.layout.SquareGraphicPosition;

import java.util.function.Consumer;

/**
 * Controle une case. Une case à différents styles (couleur et bordure)
 */
public class SquarePane extends Rectangle {
    /**
     * Les différentes couleurs et bordure possible pour la case
     */
    public enum Style {
        NORMAL,
        RED,
        BLUE
    }

    /**
     * Si la case est une case blanche ou noir (grise)
     */
    private final boolean isWhite;

    /**
     * @param isWhite       si la case est blanche
     * @param clickListener la fonction à appeler quand la case est appuyé
     * @param position      la position de la case
     * @param size the size of the square (width/height)
     */
    public SquarePane(boolean isWhite, @NotNull Consumer<SquareGraphicPosition> clickListener, @NotNull SquareGraphicPosition position, ObservableNumberValue size) {
        super();
        this.isWhite = isWhite;

        this.widthProperty().bind(size);
        this.heightProperty().bind(size);
        this.xProperty().bind(position.getX());
        this.yProperty().bind(position.getY());

        this.setOnMouseClicked(event -> clickListener.accept(position));

        setStyle(Style.NORMAL);  //Met la couleur de l'arrière plan de la case
    }

    /**
     * @param style la nouvelle couleur de l'arrière plan de l'arrière plan
     */
    public void setStyle(@NotNull Style style) {
        this.setFill(getFillColour(style));
    }

    /**
     * @param style le style de la case
     * @return la couleur de remplissage pour la case
     */
    @Contract(pure = true)
    private Paint getFillColour(@NotNull Style style) {
        switch (style) {
            case BLUE:
                return isWhite ? Color.LIGHTBLUE : Color.CORNFLOWERBLUE;
            case RED:
                return Color.PALEVIOLETRED;
            case NORMAL:
                return isWhite ? Color.WHITE : Color.LIGHTGRAY;
            default:
                throw new IllegalArgumentException("Colour de style inconnue");
        }
    }
}

File: src/main/java/ui/game/components/SquarePane.kt
package ui.game.components

import javafx.beans.value.ObservableNumberValue
import javafx.scene.paint.Color
import javafx.scene.paint.Paint
import javafx.scene.shape.Rectangle
import org.jetbrains.annotations.Contract
import ui.game.layout.SquareGraphicPosition

import java.util.function.Consumer

/**
 * Controls a square on the UI. The square can be highlighted blue or red.
 */
class SquarePane
/**
 * @param isWhite       true if the square is a white square
 * @param clickListener the method to call when the square is clicked
 * @param position      the squares position
 * @param size          the size of the square (width/height)
 */
(
        /**
         * True if the square is a white square (false if black/gray)
         */
        private val isWhite: Boolean, clickListener: Consumer<SquareGraphicPosition>, position: SquareGraphicPosition, size: ObservableNumberValue) : Rectangle() {
    /**
     * The different highlight modes
     */
    enum class Style {
        NORMAL,
        RED,
        BLUE
    }

    init {

        this.widthProperty().bind(size)
        this.heightProperty().bind(size)
        this.xProperty().bind(position.x)
        this.yProperty().bind(position.y)

        this.setOnMouseClicked { _ -> clickListener.accept(position) }

        setStyle(Style.NORMAL)  //Set the background colour
    }

    /**
     * @param style the new highlight style
     */
    fun setStyle(style: Style) {
        this.fill = getFillColour(style)
    }

    /**
     * @return the background fill colour for this highlight style
     */
    @Contract(pure = true)
    private fun getFillColour(style: Style): Paint {
        return when (style) {
            SquarePane.Style.BLUE -> if (isWhite) Color.LIGHTBLUE else Color.CORNFLOWERBLUE
            SquarePane.Style.RED -> Color.PALEVIOLETRED
            SquarePane.Style.NORMAL -> if (isWhite) Color.WHITE else Color.LIGHTGRAY
        }
    }
}
